### PR TITLE
[dev] [DeepSeek-v4] Part 4: Fusion Kernels for DSv4 Hybrid Attention

### PR DESCRIPTION
**Pull request by hxbai**
# What does this PR do ?
https://github.com/NVIDIA/cudnn-frontend release v1.24.0 added the fusion kernels for DeepSeek-V4 Hybrid Attention.

This PR integrated these kernel into MCore.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_native_parity.py [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+                    )
+                    positions = torch.arange(1, sq + 1, device=x.device).unsqueeze(1)
+                    causal_mask = (
+                        torch.where(
+                            causal_mask >= positions // self.compress_ratio, float("-inf"), 0.0
+                        )
+                        .unsqueeze(0)
+                        .expand(batch_size, -1, -1)
+                    )
+                    indexer_loss = compute_dsa_indexer_loss(
```
Comment (kunlunl) [inline tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_native_parity.py]: We shouldn't reply on mcore's implementation in the native version, it should be truly independent.
I'll push a new commit to solve this.

**Review by kunlunl — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 2]
```diff
@@ -0,0 +1,1000 @@
...
+            # Dense kernels expect K as 4-D ``(B, S_k, H_kv, D)``; our K is
+            # 3-D MQA. The unsqueeze is a free view (size-1 head dim).
+            index_score, index_lse = _compute_dense_indexer_score(
+                q_idx_bshd,
+                k_idx_bsd.unsqueeze(2),
+                w_bsh_scaled,
+                qhead_per_kv_head=idx_nh,
+                indexer_softmax_scale=indexer_softmax_scale,
+                ratio=ratio,
+            )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: fixed

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 3]
```diff
@@ -596,7 +604,7 @@ def __init__(
             softmax_scale = config.v_head_dim**-0.5
         self.softmax_scale = softmax_scale
 
-        self.force_unfused_dsa = getattr(config, 'force_unfused_dsa', True)
+        self.force_unfused_dsa = getattr(config, 'force_unfused_dsa', False)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: add a new arg to control and making the fusion as default is intentional

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 4]
```diff
@@ -0,0 +1,1000 @@
...
+    position have ``attn_l1norm <= 0`` (L1) or ``index_lse == -inf``
+    (LSE); those rows contribute 0 to the loss — the same ``row_valid``
+    semantics as the reference ``compute_dsa_indexer_loss``.
+    """
+    eps = _CLIP_PROB_MIN
+    # row_valid: rows with at least one un-masked KV position.
+    row_valid = (attn_l1norm > eps) & torch.isfinite(index_lse)
+
+    # Safe denoms: replace invalid rows with a finite value so target /
+    # log-predict don't produce NaN; the row mask zeroes their KL below.
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py:656]: The cuDNN dense score kernel writes 0.0 (not -inf) at masked positions — both SM90 and SM100 implementations do this explicitly, and the SM100 kernel comments: "Masked positions write 0 to mOut … avoids -inf-vs-empty edge case." The cuDNN test suite also asserts torch.isfinite(out).all(). So index_score is always finite here, and the eps * inf path cannot trigger.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 5]
```diff
@@ -768,7 +776,76 @@ def forward(
...
+                    )
+                else:
+                    flat_idxs, _ = build_flat_topk_idxs(
+                        window_idxs,
+                        batch_size=b, seqlen_kv=kv_full.shape[0],
+                    )
+                output = dsa_sparse_attn(
+                    query, kv_full, self.attn_sink.float(),
+                    flat_idxs, self.softmax_scale,
+                )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: added

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_native_parity.py (hunk 1 above)
Comment (hxbai) [reply tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_native_parity.py]: OK, thanks

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 6]
```diff
@@ -0,0 +1,1035 @@
...
+    positions (matching the reference's ``kl_per_element.sum(-1).mean()``).
+    """
+    eps = _CLIP_PROB_MIN
+    t = target.clamp(min=eps)
+    p = predict.clamp(min=eps)
+    kl_per_row = (t * (torch.log(t) - torch.log(p))).sum(dim=-1)  # (B, S_q)
+
+    row_valid = (topk_indices >= 0).any(dim=-1)  # (B, S_q)
+    kl_per_row = torch.where(row_valid, kl_per_row, torch.zeros_like(kl_per_row))
+    return loss_coeff * kl_per_row.mean()
```
Comment (kunlunl) [inline megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: Should we use .sum() to make it invariant when using THD/CP?

**Review by kunlunl — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py (hunk 6 above)
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: fixed by Kunlun

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/2b3c815aca37063a51ce089717f60d9adad08796

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 7]
```diff
@@ -717,7 +723,7 @@ def forward(
                         # indexer_softmax_scale; apply it here via the
                         # weights-scaling trick so the effective weights match
                         # the pre-scale-split behaviour.
-                        weights_for_unfused = weights_indexer * self.indexer.softmax_scale
+                        weights_for_unfused = weights_indexer.float() * self.indexer.softmax_scale
```
Comment (kunlunl) [inline megatron/core/transformer/experimental_attention_variant/csa.py]: Why don't pass the softmax_scale into kernel directly?

Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: This is for the unfused path. Since the fused kernel accepts the scale, we only need to multiply the scale in the unfused path.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 8]
```diff
@@ -768,7 +775,99 @@ def forward(
...
+                        sparse_loss=getattr(self.config, "dsa_indexer_use_sparse_loss", True),
+                        kv_offset=offset,
+                        calculate_per_token_loss=self.config.calculate_per_token_loss,
+                    )
+                    nvtx_range_pop("sparse_attn_kernel")
+                    if indexer_loss_coeff > 0:
+                        DSAIndexerLossLoggingHelper.save_loss_to_tracker(
+                            loss=indexer_loss,
+                            layer_number=self.layer_number,
+                            num_layers=self.config.num_layers,
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: fixed

File: megatron/training/arguments.py [hunk 9]
```diff
@@ -4601,6 +4602,13 @@ def _add_experimental_attention_variant_args(parser):
         'transformer layer (valid values: 0, 4, 128). '
         'The list length must equal num_layers.',
     )
+    group.add_argument(
+        '--no-dsa-kernel-fusion',
+        action='store_false',
+        help='Disable fused DSA sparse-attention kernels (FlashMLA + cuDNN DSA) '
+        'and fall back to unfused PyTorch implementations.',
+        dest='apply_dsa_kernel_fusion',
+    )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/training/arguments.py:4611]: The review comment is not correct. For `store_false`, if this arg is not set, the default value of `apply_dsa_kernel_fusion` will be True. This is what we need.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 10]
```diff
@@ -768,7 +775,99 @@ def forward(
...
+                if self.training and torch.is_grad_enabled():
+                    # Path B: fused indexer (with loss) + sparse attn
+                    nvtx_range_push("compressed_indices")
+                    x_det = x.detach()
+                    qr_det = qr.detach()
+                    q_indexer, k_indexer, weights_indexer = self.indexer.forward_before_topk(
+                        x_det, qr_det, packed_seq_params
+                    )
+                    nvtx_range_pop("compressed_indices")
+                    indexer_loss_coeff = getattr(self.config, 'dsa_indexer_loss_coeff', 0.0)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: fixed

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/d91242ade1a1e5bb571d6efbbd2ffe2c66128aa7

File: megatron/training/arguments.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/training/arguments.py:4611]: see https://docs.python.org/3/library/argparse.html#action 

'store_true' and 'store_false' - These are special cases of 'store_const' that respectively store the values True and False with default values of False and True:

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/8ba9cc0dbe05f295be479fa036fd0f6947be5b20

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/fca33ac3c6b17fabc543803905421c37ff7a37e3

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/b86e4c51097eec33d998e147cfb0891b2774981e

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/30a94d9a09a6d2d9c811917e7dc6426d1e2cca42

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 11]
```diff
@@ -768,7 +775,99 @@ def forward(
             nvtx_range_pop("sparse_attn_kernel")
 
         else:
-            raise ValueError("Fused path is not supported for CompressedSparseAttention")
+            if self.compress_ratio > 1 and n_compressed > 0 and self.indexer is not None:
```
Comment (Victarry) [inline megatron/core/transformer/experimental_attention_variant/csa.py]: **[SUGGESTION Refactor]** Consider splitting the CSA forward path into smaller private helpers.

`CompressedSparseAttention.forward()` currently mixes KV construction, unfused fallback logic, fused Path A/B/C dispatch, indexer-loss logging, and loss attachment in one large nested function. This makes it hard to reason about which path is responsible for which part of the fused kernel behavior.

One possible structure is to keep `forward()` as the high-level flow and move the path-specific logic into private helpers, for example:

- `_build_kv_full()` for original plus compressed KV construction
- `_forward_unfused_csa()` for the PyTorch fallback path
- `_forward_fused_static_indices()` for Path A, where the fused attention consumes window or deterministic compressed indices
- `_forward_fused_indexer_inference()` for Path C
- `_forward_fused_indexer_training()` for Path B, including fused indexer loss

The fused Path A/B/C dispatch can still stay expanded in `forward()` so the main control flow remains visible, while the detailed kernel plumbing is isolated in focused helpers.

**Review by Victarry — APPROVED** [review]

File: megatron/core/transformer/experimental_attention_variant/csa.py (hunk 11 above)
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: refactored

**Review by kunlunl — APPROVED** [review]

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/9d74bc615e516d6f9a346d616b510c64c31a1c7d

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/00a877f3e8b9a4bb86e44ea7e9931bb37639a616

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/a8646909cbb8b4b9b2e4c9563d2d1ca382841ba4

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/70a667395bc6ea68d1ca1d66dd03e829a8fc33fd

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/a31271ed5081026b50b083fd09ef68761c150938

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/4e560b77b541b9d6a918ba6829e1243fb609e1c8

File: pyproject.toml [hunk 12]
```diff
@@ -98,6 +98,7 @@ dev = [
     "megatron-energon[av_decode]~=6.0",
     "av",
     "flashinfer-python>=0.5.0,<0.7.0",
+    "nvidia-cudnn-frontend==1.24.0",
```
Comment (ko3n1g) [inline pyproject.toml]: do we really need such a strict pin for all of MLM? Can we have a version guard inside core instead?

**Review by ko3n1g — CHANGES_REQUESTED** [review]
question regarding pin strictness

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/04bade08b2ff40eebb24c305298ef72871c564f3

File: pyproject.toml (hunk 12 above)
Comment (hxbai) [reply pyproject.toml]: removed the pin

**Review by ko3n1g — APPROVED** [review]
ty!

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/ccc8449f554142e5ccd1745bd2004c2fe9027c5d

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4894/commits/5a806d1807d25be4d9ad3bfe06c047671e319f44
